src/coreComponents/constitutive/solid/DruckerPrager.hpp

```diff
@@ -0,0 +1,449 @@
+
+  // the return mapping can in general be written as a newton iteration.
+  // here we have a linear problem, so the algorithm will converge in one
+  // iteration, but this is a template for more general models with either
+  // nonlinear hardening or yield surfaces.
+
+  real64 solution[3], residual[3], delta[3];
+  real64 jacobian[3][3] = {{}}, jacobianInv[3][3] = {{}};
+
+  solution[0] = trialP; // initial guess for newP
+  solution[1] = trialQ; // initial guess for newQ
+  solution[2] = 1e-5;   // initial guess for plastic multiplier
+
+  real64 norm, normZero = 1e30;
+
+  // begin newton loop
+
+  for( localIndex iter=0; iter<20; ++iter )
+  {
+    // apply a linear cohesion decay model,
+    // then check for complete cohesion loss
+
+    m_newCohesion[k][q] = m_oldCohesion[k][q] + solution[2] * m_hardening[k];
+    real64 cohesionDeriv = m_hardening[k];
+
+    if( m_newCohesion[k][q] < 0 )
+    {
+      m_newCohesion[k][q] = 0;
+      cohesionDeriv = 0;
+    }
+
+    // assemble residual system
+    // resid1 = P - trialP + dlambda*bulkMod*dG/dP = 0
+    // resid2 = Q - trialQ + dlambda*3*shearMod*dG/dQ = 0
+    // resid3 = F = 0
+
+    residual[0] = solution[0] - trialP + solution[2] * m_bulkModulus[k] * m_dilation[k];
+    residual[1] = solution[1] - trialQ + solution[2] * 3 * m_shearModulus[k];
+    residual[2] = solution[1] + m_friction[k] * solution[0] - m_newCohesion[k][q];
+
+    // check for convergence
+
+    norm = LvArray::tensorOps::l2Norm< 3 >( residual );
+
+    if( iter==0 )
+    {
+      normZero = norm;
+    }
+
+    //printf("%ld %.2e\n",iter,norm);
+
+    if( norm < 1e-8*(normZero+1))
+    {
+      break;
+    }
+
+    // solve Newton system
+
+    jacobian[0][0] = 1;
+    jacobian[0][2] = m_bulkModulus[k] * m_dilation[k];
+    jacobian[1][1] = 1;
+    jacobian[1][2] = 3 * m_shearModulus[k];
+    jacobian[2][0] = m_friction[k];
+    jacobian[2][1] = 1;
+    jacobian[2][2] = -cohesionDeriv;
+
+    LvArray::tensorOps::invert< 3 >( jacobianInv, jacobian );
+    LvArray::tensorOps::Ri_eq_AijBj< 3, 3 >( delta, jacobianInv, residual );
+
+    for( localIndex i=0; i<3; ++i )
+    {
+      solution[i] -= delta[i];
+    }
+  }
+
+  // re-construct stress = P*eye + sqrt(2/3)*Q*nhat
+
+  twoInvariant::stressRecomposition( solution[0],
+                                     solution[1],
+                                     deviator,
+                                     stress );
+
+  // construct consistent tangent stiffness
+  // note: if trialQ = 0, we will get a divide by zero error below,
+  // but this is an unphysical (zero-strength) state anyway
+
+  LvArray::tensorOps::fill< 6, 6 >( stiffness, 0 );
+
+  real64 c1 = 2 * m_shearModulus[k] * solution[1] / trialQ;
+  real64 c2 = jacobianInv[0][0] * m_bulkModulus[k] - c1 / 3;
+  real64 c3 = sqrt( 2./3 ) * 3 * m_shearModulus[k] * jacobianInv[0][1];
+  real64 c4 = sqrt( 2./3 ) * m_bulkModulus[k] * jacobianInv[1][0];
```

Comment:
We may write the double values in a consistent way. Here we can find       2.       3      1.0     etc.